File: ADT/DSU/1319.py
class Solution:
    def __init__(self):
        self.parent = []
        self.rank = []

    def find(self, x: int) -> int:
        if x == self.parent[x]:
            return x
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def Union(self, x: int, y: int) -> None:
        x_parent = self.find(x)
        y_parent = self.find(y)

        if x_parent == y_parent:
            return

        if self.rank[x_parent] > self.rank[y_parent]:
            self.parent[y_parent] = x_parent
        elif self.rank[x_parent] < self.rank[y_parent]:
            self.parent[x_parent] = y_parent
        else:
            self.parent[x_parent] = y_parent
            self.rank[y_parent] += 1

    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n - 1:
            return -1

        self.parent = list(range(n))
        self.rank = [0] * n

        components = n
        for vec in connections:
            if self.find(vec[0]) != self.find(vec[1]):
                components -= 1
                self.Union(vec[0], vec[1])

        return components - 1

File: ADT/DSU/1319.js
/**
 * @param {number} n
 * @param {number[][]} connections
 * @return {number}
 */
var makeConnected = function (n, connections) {
    const parent = new Array(n).fill(0).map((_, i) => i)
    const rank = new Array(n).fill(1)

    const find = (x) => {
        if (parent[x] !== x) {
            parent[x] = find(parent[x])
        }
        return parent[x]
    }

    const Union = (x, y) => {
        const px = find(x)
        const py = find(y)
        if (px !== py) {
            if (rank[px] > rank[py]) {
                parent[py] = px
            } else if (rank[px] < rank[py]) {
                parent[px] = py
            } else {
                parent[px] = py
                rank[py] += 1
            }
        }
    }

    if (connections.length < n - 1) {
        return -1
    }
    let components = n
    for (const vec of connections) {
        if (find(vec[0]) !== find(vec[1])) {
            components--
            Union(vec[0], vec[1])
        }
    }
    return components - 1
};